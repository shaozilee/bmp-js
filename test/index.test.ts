import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';

import bmp from '../lib/';

const createPath = (p: string) => path.join(process.cwd(), p);
const readFile = (p: string) => fs.readFileSync(createPath(p));

const checksum = (str: Buffer, algorithm = 'md5', encoding = 'hex') =>
  crypto
    .createHash(algorithm)
    .update(str.toString(), 'utf8')
    // @ts-ignore
    .digest(encoding);

describe('decode', () => {
  const decodeTest = (bitPP: number | string, options = {}) => () => {
    const buff = readFile(`./test/images/bit${bitPP}.bmp`);
    const { data } = bmp.decode(buff, options);
    expect(checksum(data)).toMatchSnapshot();
  };

  test('errors for non bmp files', () => {
    const buff = readFile('package.json');
    expect(() => bmp.decode(buff)).toThrow();
  });

  test('1-bit', decodeTest(1));

  describe('4-bit', () => {
    test('normal', decodeTest(4));
    test('RLE', decodeTest('4_RLE'));
  });

  describe('8-bit', () => {
    test('normal', decodeTest(8));
    test('RLE', decodeTest('8_RLE'));
  });

  describe('16-bit', () => {
    test('565', decodeTest('16_565'));
    test('a444', decodeTest('16_a444'));
    test('a555', decodeTest('16_a555'));
    test('x444', decodeTest('16_x444'));
    test('x555', decodeTest('16_x555'));
  });

  test('24-bit', decodeTest(24));

  describe('32-bit', () => {
    test('normal', decodeTest(32));
    test('alpha', decodeTest('32_alpha'));
    test('alpha - toRGBA', decodeTest('32_alpha', { toRGBA: true }));
  });
});

describe('encode', () => {
  const encodeTest = (bitPP: number, file = 32) => () => {
    const buff = readFile(`./test/images/bit${file}.bmp`);
    const bitmap = bmp.decode(buff);

    bitmap.bitPP = bitPP;

    const { data } = bmp.encode(bitmap);
    expect(checksum(data)).toMatchSnapshot();
  };

  const errorTest = (bitPP: number) => () => {
    const buff = readFile('./test/images/bit32.bmp');
    const bitmap = bmp.decode(buff);

    bitmap.bitPP = bitPP;

    expect(() => bmp.encode(bitmap)).toThrow();
  };

  test('1-bit', encodeTest(1));

  describe('4-bit', () => {
    test('errors without color pallette', errorTest(4));
    test('works when colors provided', encodeTest(4, 4));
  });

  describe('8-bit', () => {
    test('errors without color pallette', errorTest(8));
    test('works when colors provided', encodeTest(8, 8));
  });

  test('16-bit', encodeTest(16));
  test('24-bit', encodeTest(24));
  test('32-bit', encodeTest(32));
});

describe('decode -> encode', () => {
  const compareDecodeEncode = (
    bitPP: number,
    file: number | string = bitPP
  ) => () => {
    const before = readFile(`./test/images/bit${file}.bmp`);
    const decoded = bmp.decode(before);

    decoded.bitPP = bitPP;

    const after = bmp.encode(decoded).data;

    expect(checksum(before)).toBe(checksum(after));
  };

  test('1-bit', compareDecodeEncode(1));
  test('4-bit', compareDecodeEncode(4));
  test('8-bit', compareDecodeEncode(8));
  test('16-bit', compareDecodeEncode(16, '16_x555'));
  test('24-bit', compareDecodeEncode(24));
  test('32-bit', compareDecodeEncode(32));
});
