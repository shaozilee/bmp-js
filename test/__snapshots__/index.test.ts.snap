// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode 1-bit 1`] = `"cec76def32993b226b554d08f78948d9"`;

exports[`decode 4-bit RLE 1`] = `"beac91cf64120346736c9b1383f9da60"`;

exports[`decode 4-bit normal 1`] = `"df219955a66ea20a8ccb4f0537930458"`;

exports[`decode 8-bit RLE 1`] = `"81bbf4f1f828dde5665bf1d798e39e72"`;

exports[`decode 8-bit normal 1`] = `"6697bb6ab8078d739fc40bbe558bf4ce"`;

exports[`decode 16-bit 565 1`] = `"aa2751f1e5c305448a9622a926ca3628"`;

exports[`decode 16-bit a444 1`] = `"e16a3905a7a21a0e8f504285fccb3544"`;

exports[`decode 16-bit a555 1`] = `"6fb3e1971fa948d1cc1fe268a037f05a"`;

exports[`decode 16-bit x444 1`] = `"988a4d08bdaf12a32954e81f5794503a"`;

exports[`decode 16-bit x555 1`] = `"fea6f96b3099726499275ae8b395df02"`;

exports[`decode 24-bit 1`] = `"389fa2b1873b5aa14f7ecab58fe96bd9"`;

exports[`decode 32-bit alpha - toRGBA 1`] = `"1fb6cd994623e11eff63c44693a405ad"`;

exports[`decode 32-bit alpha 1`] = `"5cb1c7699a421a7bead9adc527e9da8f"`;

exports[`decode 32-bit normal 1`] = `"5cb1c7699a421a7bead9adc527e9da8f"`;

exports[`encode 1-bit 1`] = `"083203daf9d76316bd464783281ba8f8"`;

exports[`encode 4-bit works when colors provided 1`] = `"1299bf1998cd01a99c0c8bc671526cf9"`;

exports[`encode 8-bit works when colors provided 1`] = `"b55da02d398a62d53a16885d86d343c0"`;

exports[`encode 16-bit 1`] = `"f4177dbf3979ec23994641637a767347"`;

exports[`encode 24-bit 1`] = `"6930e0e0e345d5ad834817260dc1c00c"`;

exports[`encode 32-bit 1`] = `"541e029c95f3466622595433a03eaa3e"`;
